File: 2024/09/Node/main.js
'use strict';

const fs = require('fs');

function computeDiskChecksum() {
  const input = fs.readFileSync('../input.txt', 'utf-8').trim();
  let processingFile = true;
  let currentFileIndex = 0;
  const diskLayout = [];

  for (const size of input) {
    const blockSize = Number(size);
    if (processingFile) {
      diskLayout.push(...Array(blockSize).fill(currentFileIndex));
      currentFileIndex++;
    } else {
      diskLayout.push(...Array(blockSize).fill('.'));
    }
    processingFile = !processingFile;
  }

  for (let i = 0; i < diskLayout.length; i++) {
    if (diskLayout[i] === '.') {
      for (let j = diskLayout.length - 1; j > i; j--) {
        if (diskLayout[j] !== '.') {
          diskLayout[i] = diskLayout[j];
          diskLayout[j] = '.';
          break;
        }
      }
    }
  }

  const totalChecksum = diskLayout.reduce((sum, block, position) => {
    return block !== '.' ? sum + block * position : sum;
  }, 0);

  console.log('Checksum:', totalChecksum);
}

computeDiskChecksum();

File: 2024/09/Go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func computeDiskChecksum() {
	file, err := os.Open("../input.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Scan()
	input := scanner.Text()

	processingFile := true
	currentFileIndex := 0
	diskLayout := []interface{}{} 

	for _, sizeChar := range input {
		blockSize, err := strconv.Atoi(string(sizeChar))
		if err != nil {
			fmt.Println("Error converting block size:", err)
			return
		}

		if processingFile {
			for i := 0; i < blockSize; i++ {
				diskLayout = append(diskLayout, currentFileIndex)
			}
			currentFileIndex++
		} else {
			for i := 0; i < blockSize; i++ {
				diskLayout = append(diskLayout, ".")
			}
		}
		processingFile = !processingFile
	}
	for i := 0; i < len(diskLayout); i++ {
		if diskLayout[i] == "." {
			for j := len(diskLayout) - 1; j > i; j-- {
				if diskLayout[j] != "." {
					diskLayout[i] = diskLayout[j]
					diskLayout[j] = "."
					break
				}
			}
		}
	}
	checksum := 0
	for i, block := range diskLayout {
		if block != "." {
			checksum += block.(int) * i
		}
	}
	fmt.Println("Checksum:", checksum)
}

func main() {
	computeDiskChecksum()
}
